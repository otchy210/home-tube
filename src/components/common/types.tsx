export type HTMLElementProps = React.HTMLAttributes<HTMLElement>;

export type HTMLDivProps = React.HTMLAttributes<HTMLDivElement>;

export type HTMLButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement>;

export type HTMLFormProps = React.FormHTMLAttributes<HTMLFormElement>;

export type HTMLLabelProps = React.LabelHTMLAttributes<HTMLLabelElement>;

export type HTMLInputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type'> & {
    isInvalid?: boolean;
};

export type HTMLCheckboxProps = HTMLInputProps & {
    label?: string;
};

export type HTMLSelectProps = React.SelectHTMLAttributes<HTMLSelectElement>;
