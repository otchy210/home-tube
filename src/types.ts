export type VideoViewMode = 'normal' | 'theater' | 'fullScreen';

export interface FolderData {
    path: string;
    name: string;
    folders: FolderData[];
}

export interface TreeNode {
    key: string;
    title: string;
    children?: TreeNode[];
}
