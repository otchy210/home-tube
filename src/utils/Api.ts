import {
    AllTags,
    AppConfig,
    AppConfigValidationError,
    Json,
    ServerStatus,
    VideoConverterStatus,
    VideoDetails,
    VideoDocument,
    VideoProperties,
    VideoValues,
} from '@otchy/home-tube-api/dist/types';
import { createSearchParams } from 'react-router-dom';
import TimeSizeLimitedCache from './TimeSizeLimitedCache';

type Method = 'GET' | 'POST' | 'DELETE';

type FetchOptions = {
    method: Method;
    headers?: { [name: string]: string };
    body?: string;
};

type Params = Record<string, string | string[]>;

type ApiSafeParams = Record<string, string>;

export class Api {
    private apiHost;
    private detailsCache;
    constructor(apiHost: string) {
        this.apiHost = apiHost;
        this.detailsCache = new TimeSizeLimitedCache<string, VideoDetails>(1000 * 60 * 10, 128);
    }
    private getApiUrl(apiPath: string, params?: Params): string {
        if (!params) {
            return `${this.apiHost}${apiPath}`;
        }
        const apiSafeParams = Object.entries(params).reduce((map, [name, value]) => {
            if (!value) {
                return map;
            }
            if (Array.isArray(value)) {
                const filteredValue = value.filter((item) => {
                    return item.trim().length > 0;
                });
                if (filteredValue.length > 0) {
                    map[name] = JSON.stringify(filteredValue);
                }
            } else {
                map[name] = value;
            }
            return map;
        }, {} as ApiSafeParams);
        if (Object.keys(apiSafeParams).length === 0) {
            return `${this.apiHost}${apiPath}`;
        }
        const searchParams = createSearchParams(apiSafeParams);
        return `${this.apiHost}${apiPath}?${searchParams}`;
    }
    private call<T>(method: Method, apiPath: string, options: { body?: Json; params?: Params } = {}): Promise<T> {
        const { body, params } = options;
        const apiUrl = this.getApiUrl(apiPath, params);
        const fetchOptions: FetchOptions = { method };
        if (body) {
            fetchOptions.headers = {
                'Content-Type': 'application/json',
            };
            fetchOptions.body = JSON.stringify(body);
        }
        return new Promise((resolve, reject) => {
            fetch(apiUrl, fetchOptions)
                .then((response) => {
                    if (!response.ok) {
                        const reason = `Api "${apiPath}" doesn't return 200.`;
                        console.error(reason, response);
                        reject(reason);
                    } else {
                        resolve(response.json());
                    }
                })
                .catch((e) => {
                    const reason = `Api "${apiPath}" failed.`;
                    console.error(reason, e);
                    reject(reason);
                });
        });
    }
    private get<T>(apiPath: string, params?: Record<string, string | string[]>): Promise<T> {
        return this.call('GET', apiPath, { params });
    }
    private post<T>(apiPath: string, body: Json, params?: Record<string, string | string[]>): Promise<T> {
        return this.call('POST', apiPath, { body, params });
    }
    private delete<T>(apiPath: string, params?: Record<string, string | string[]>): Promise<T> {
        return this.call('DELETE', apiPath, { params });
    }
    getAppConfig(): Promise<AppConfig> {
        return this.get<AppConfig>('/appConfig');
    }
    postAppConfig(appConfig: AppConfig): Promise<AppConfig | AppConfigValidationError[]> {
        return this.post<AppConfig>('/appConfig', appConfig);
    }
    getServerStatus(): Promise<ServerStatus> {
        return this.get('/serverStatus');
    }
    search(params?: Record<string, string | string[]>): Promise<Set<VideoDocument>> {
        return this.get<Set<VideoDocument>>('/search', params);
    }
    getDetails(key: string): Promise<VideoDetails> {
        const cachedDetails = this.detailsCache.get(key);
        if (cachedDetails) {
            return Promise.resolve(cachedDetails);
        }
        return new Promise((resolve, reject) => {
            this.get<VideoDetails>('/details', { key })
                .then((videoDetails) => {
                    this.detailsCache.add(key, videoDetails);
                    resolve(videoDetails);
                })
                .catch((e) => {
                    reject(e);
                });
        });
    }
    getThumbnailsUrl(key: string, minute: string): string {
        return this.getApiUrl('/thumbnails', { key, minute });
    }
    getSnapshotUrl(key: string): string {
        return this.getApiUrl('/snapshot', { key });
    }
    postSnapshot(key: string, dataURL: string): Promise<void> {
        return this.post<void>('/snapshot', { dataURL }, { key });
    }
    getVideoUrl(key: string): string {
        return this.getApiUrl('/video', { key });
    }
    postProperties(key: string, properties: VideoProperties): Promise<VideoProperties> {
        this.detailsCache.remove(key);
        return this.post<VideoProperties>('/properties', properties, { key });
    }
    postConvert(key: string, type: string): Promise<{ status: VideoConverterStatus }> {
        this.detailsCache.remove(key);
        return this.post<{ status: VideoConverterStatus }>('/convert', null, { key, type });
    }
    deleteConvert(key: string, type: string): Promise<{ status: VideoConverterStatus }> {
        this.detailsCache.remove(key);
        return this.delete<{ status: VideoConverterStatus }>('/convert', { key, type });
    }
    getAllTags(): Promise<AllTags> {
        return this.get<AllTags>('/allTags');
    }
    postRename(key: string, name: string): Promise<VideoValues> {
        return this.post<VideoValues>('/rename', {}, { key, name });
    }
}
