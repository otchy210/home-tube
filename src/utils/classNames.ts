type Arg = undefined | null | string | Arg[];

class ClassNames {
    private classNames = new Set<string>();

    constructor(...args: Arg[]) {
        this.addArray(args);
    }

    private addSingle(arg: string) {
        const className = arg.trim();
        if (className.length > 0) {
            this.classNames.add(className);
        }
    }

    private addString(arg: string) {
        const args = arg.split(/\s+/);
        if (args.length === 1) {
            this.addSingle(args[0]);
        } else if (args.length > 1) {
            args.forEach((arg) => {
                this.addSingle(arg);
            });
        }
    }

    private addArray(args: Arg[]) {
        args.forEach((arg) => {
            if (Array.isArray(arg)) {
                this.addArray(arg);
            } else if (typeof arg === 'string') {
                this.addString(arg);
            }
        });
    }

    add(...args: Arg[]) {
        this.addArray(args);
        return this;
    }

    build() {
        if (this.classNames.size === 0) {
            return null;
        }
        return Array.from(this.classNames).sort().join(' ');
    }
}

export const classNames = (...args: Arg[]) => {
    return new ClassNames(args);
};
