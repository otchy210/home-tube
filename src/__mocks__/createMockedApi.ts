import { Api } from '../utils/Api';

const createMockedApi = () => {
    const api = new Api('http://dummy-api-host');
    jest.spyOn(api, 'getAppConfig');
    jest.spyOn(api, 'getSnapshotUrl');
    jest.spyOn(api, 'getServerStatus');
    return api;
};

export default createMockedApi;
