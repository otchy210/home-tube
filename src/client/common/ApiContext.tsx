import { createContext } from 'react';
import { Api } from './Api';

const ApiContext = createContext<Api>(new Api());

export default ApiContext;
