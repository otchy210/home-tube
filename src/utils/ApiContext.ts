import { createContext, useContext } from 'react';
import { Api } from './Api';

export const ApiContext = createContext<Api | undefined>(undefined);

export const useApi = (): Api => {
    const api = useContext(ApiContext);
    if (!api) {
        throw new Error('Api is not provided via context');
    }
    return api;
};
