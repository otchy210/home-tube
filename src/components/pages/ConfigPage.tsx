import { AppConfig, AppConfigValidationError, ServerStatus, Storage } from '@otchy/home-tube-api/dist/types';
import { TFunction } from 'i18next';
import React, { useEffect, useState } from 'react';
import { Form } from 'react-bootstrap';
import styled from 'styled-components';
import { useSetTitle } from '../../hooks/useSetTitle';
import Reload from '../../images/reload.svg';
import { getAppConfigDeepCopy } from '../../utils/ObjectUtils';
import { waitFor } from '../../utils/TimerUtils';
import Spinner from '../atoms/Spinner';
import { Badge } from '../common/badges';
import { DangerButton, LinkButton, SecondaryButton, SubmitButton } from '../common/buttons';
import { FormTextInput } from '../common/form';
import { HorizontalStack, VerticalStack } from '../common/layouts';
import { Row, Col } from '../common/layouts';
import DelayedSpinner from '../molecules/DelayedSpinner';
import ServerStatusProperty from '../molecules/ServerStatusProperty';
import { useApi } from '../providers/ApiProvider';
import { useI18n } from '../providers/I18nProvider';
import { useToast } from '../providers/ToastsProvider';

/*
Translations for type StorageMonitorStatus = 'initialized' | 'reading' | 'waiting' | 'stopped';
t('initialized') t('reading') t('waiting') t('stopped')
*/

/*
Translations for updating AppConfig error messages
t("Storage doesn't exist") t("ffmpeg command doesn't exist")
*/

const Title = styled.p.attrs({ className: 'h1 pt-3' })``;

const PropertyTitle = styled.p.attrs({ className: 'h5 mt-3' })``;

const ReloadIcon = styled(Reload).attrs({ width: 32, height: 32 })`
    cursor: pointer;
`;

const isAppConfigValidationErrors = (results: AppConfig | AppConfigValidationError[]): results is AppConfigValidationError[] => {
    return Array.isArray(results);
};

type StorageValidatinErrors = Map<number, string>;

export const validateStorages = (storages: Storage[], t: TFunction): StorageValidatinErrors => {
    const errors = new Map<number, string>();
    const seenPaths: { [path: string]: number[] } = {};
    storages.forEach((storage, index) => {
        const { path } = storage;
        if (path.length === 0) {
            errors.set(index, t('Path is empty.'));
            return;
        }
        if (seenPaths[path] === undefined) {
            seenPaths[path] = [];
        }
        seenPaths[path].push(index);
    });
    Object.values(seenPaths)
        .filter((indexes) => indexes.length > 1)
        .forEach((indexes) => {
            indexes.forEach((index) => {
                errors.set(index, t('Path is duplicated.'));
            });
        });
    return errors;
};

const ConfigPage: React.FC = () => {
    const [appConfig, setAppConfig] = useState<AppConfig>();
    const [serverStatus, setServerStatus] = useState<ServerStatus>();
    const [updated, setUpdated] = useState<boolean>(false);
    const [hasError, setHasError] = useState<boolean>(false);
    const [submitting, setSubmitting] = useState<boolean>(false);
    const [storageValidationErrors, setStorageValidationErrors] = useState<StorageValidatinErrors>(new Map<number, string>());
    const { t } = useI18n();
    const api = useApi();
    const toast = useToast();
    useSetTitle(t('Config'));
    const loadAppConfig = () => {
        api.getAppConfig()
            .then(setAppConfig)
            .catch((e) => {
                console.error(e);
                toast.addError(t('Config page'), t('Failed to load.'));
                setHasError(true);
            });
        setUpdated(false);
        setStorageValidationErrors(new Map<number, string>());
    };
    const loadServerStatus = () => {
        setServerStatus(undefined);
        Promise.all([api.getServerStatus(), waitFor(500)]).then(([serverStatus]) => {
            setServerStatus(serverStatus);
        });
    };
    useEffect(() => {
        loadAppConfig();
        loadServerStatus();
    }, []);
    if (!appConfig) {
        return (
            <>
                <Title>{t('Config')}</Title>
                <p>{!hasError && <DelayedSpinner />}</p>
            </>
        );
    }
    const updateStorageValue = (i: number, updater: (storage: Storage) => void) => {
        const copy = getAppConfigDeepCopy(appConfig);
        if (copy.storages.length < i + 1) {
            return;
        }
        updater(copy.storages[i]);
        setAppConfig(copy);
        setUpdated(true);
    };
    const updateStoragePath = (i: number, value: string) => {
        updateStorageValue(i, (storage) => {
            storage.path = value.trim();
        });
        setUpdated(true);
    };
    const updateStorageEnabled = (i: number, value: boolean) => {
        updateStorageValue(i, (storage) => {
            storage.enabled = value;
        });
        setUpdated(true);
    };
    const removeStorage = (i: number) => {
        const copy = getAppConfigDeepCopy(appConfig);
        if (copy.storages.length < i + 1) {
            return;
        }
        copy.storages.splice(i, 1);
        setAppConfig(copy);
        setUpdated(true);
    };
    const addStorage = () => {
        const copy = getAppConfigDeepCopy(appConfig);
        copy.storages.push({
            path: '',
            enabled: true,
        });
        setAppConfig(copy);
        setUpdated(true);
    };
    const updateFfmpeg = (value: string) => {
        const copy = getAppConfigDeepCopy(appConfig);
        copy.ffmpeg = value.trim();
        setAppConfig(copy);
        setUpdated(true);
    };
    const trySubmit = () => {
        const storageValidationErrors = validateStorages(appConfig.storages, t);
        setStorageValidationErrors(storageValidationErrors);
        if (storageValidationErrors.size > 0) {
            return;
        }
        setSubmitting(true);
        api.postAppConfig(appConfig)
            .then((results) => {
                if (isAppConfigValidationErrors(results)) {
                    toast.addError(
                        t('Config page'),
                        results.map((result) => {
                            return `${t(result.message)}: ${result.source}`;
                        })
                    );
                } else {
                    toast.addSuccess(t('Config page'), t('Updated successfully.'));
                    setUpdated(false);
                    loadServerStatus();
                }
            })
            .catch((e) => {
                console.error(e);
                toast.addError(t('Config page'), t('Failed to update.'));
            })
            .finally(() => {
                setSubmitting(false);
            });
    };
    return (
        <>
            <Row>
                <Col width={[12, 12, 12, 7]}>
                    <Title>{t('Config')}</Title>
                    <Form>
                        <PropertyTitle>{t('Video storage path')}</PropertyTitle>
                        <Form.Text className="text-muted">{t('Add your video storage path which has your videos.')}</Form.Text>
                        {appConfig.storages.map((storage, i) => {
                            return (
                                <VerticalStack key={`storage-${i}`}>
                                    <HorizontalStack gap={2} className="mt-1">
                                        <FormTextInput
                                            value={storage.path}
                                            isInvalid={storageValidationErrors.get(i) !== undefined}
                                            onChange={(e) => {
                                                updateStoragePath(i, e.target.value);
                                            }}
                                            data-testid={`storage-${i}-path`}
                                        />
                                        <Form.Check
                                            type="checkbox"
                                            className="text-nowrap"
                                            label={t('enabled')}
                                            checked={storage.enabled}
                                            onChange={(e) => {
                                                updateStorageEnabled(i, e.target.checked);
                                            }}
                                            id={`storage-${i}-enabled`}
                                            data-testid={`storage-${i}-enabled`}
                                        />
                                        <DangerButton
                                            onClick={() => {
                                                removeStorage(i);
                                            }}
                                            data-testid={`storage-${i}-delete`}
                                        >
                                            {t('Delete')}
                                        </DangerButton>
                                    </HorizontalStack>
                                    {storageValidationErrors.has(i) ? <Form.Text className="text-danger">{storageValidationErrors.get(i)}</Form.Text> : null}
                                </VerticalStack>
                            );
                        })}
                        <HorizontalStack gap={2} className="mt-1">
                            <SecondaryButton onClick={addStorage} data-testid={`add-storage`}>
                                {t('Add video storage path')}
                            </SecondaryButton>
                        </HorizontalStack>

                        <PropertyTitle>{t('ffmpeg path')}</PropertyTitle>
                        <Form.Text className="text-muted">{t('No need to set unless you want to change it from default.')}</Form.Text>
                        <FormTextInput
                            value={appConfig.ffmpeg}
                            className="mt-1"
                            onChange={(e) => {
                                updateFfmpeg(e.target.value);
                            }}
                            data-testid={`ffmpeg`}
                        />

                        <HorizontalStack gap={2} className="mt-3">
                            <SubmitButton disabled={!updated} submitting={submitting} data-testid="save-config" onClick={trySubmit}>
                                {t('Save config')}
                            </SubmitButton>
                            <LinkButton
                                onClick={() => {
                                    loadAppConfig();
                                }}
                                disabled={!updated || submitting}
                                data-testid="cancel"
                            >
                                {t('Cancel')}
                            </LinkButton>
                        </HorizontalStack>
                    </Form>{' '}
                </Col>
                <Col width={[12, 12, 12, 5]} className="ps-lg-5">
                    {' '}
                    <Title>
                        {t('Server status')} <ReloadIcon onClick={loadServerStatus} />
                    </Title>
                    {!serverStatus && <Spinner />}
                    {serverStatus && (
                        <>
                            <PropertyTitle>{t('Searchable videos')}</PropertyTitle>
                            <p>{serverStatus.indexedVideo}</p>
                            <PropertyTitle>{t('Storages')}</PropertyTitle>
                            {Object.entries(serverStatus.storages).map(([path, info]) => {
                                const badgeVariant = (() => {
                                    switch (info.status) {
                                        case 'initialized':
                                            return 'primary';
                                        case 'waiting':
                                            return 'secondary';
                                        case 'reading':
                                            return 'warning';
                                        case 'stopped':
                                            return 'danger';
                                    }
                                })();
                                return (
                                    <p key={`storage-${path}`}>
                                        <b>{path}</b>
                                        <br />
                                        <Badge variant={badgeVariant}>{t(info.status)}</Badge>{' '}
                                        {info.size > 0 ? t('{{count}} movies found.', { count: info.size }) : t('Movies not found.')}
                                        <br />
                                    </p>
                                );
                            })}
                            <ServerStatusProperty title={t('Meta data')} status={serverStatus.meta} />
                            <ServerStatusProperty title={t('Thumbnails')} status={serverStatus.thumbnails} />
                            <ServerStatusProperty title={t('Snapshot')} status={serverStatus.snapshot} />
                            <ServerStatusProperty title={t('Recommended MP4')} status={serverStatus.mp4} />
                        </>
                    )}
                </Col>
            </Row>
        </>
    );
};

export default ConfigPage;
