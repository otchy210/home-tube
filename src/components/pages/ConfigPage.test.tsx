import React from 'react';
import { fireEvent, render, RenderResult, waitFor } from '@testing-library/react';
import ConfigPage from './ConfigPage';
import createMockedApi from '../../__mocks__/createMockedApi';
import { ApiContext } from '../../utils/ApiContext';
import { AppConfig } from '@otchy/home-tube-api/dist/types';

describe('ConfigPage', () => {
    const mockedApi = createMockedApi();
    const mockedApiGetAppConfig = mockedApi.getAppConfig as jest.Mock;

    const testWithAppConfig = async (appConfig: AppConfig, test: (renderResult: RenderResult) => Promise<void>, done: jest.DoneCallback) => {
        mockedApiGetAppConfig.mockReturnValue(
            new Promise((resolve) => {
                resolve(appConfig);
            })
        );
        (async () => {
            const renderResult = render(
                <ApiContext.Provider value={mockedApi}>
                    <ConfigPage />
                </ApiContext.Provider>
            );
            await waitFor(() => renderResult.queryByTestId('save-config') !== null);
            await test(renderResult);
            done();
        })();
    };

    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('renders initial state properly', (done) => {
        testWithAppConfig(
            { storages: [] },
            async ({ container }) => {
                expect(container).toMatchSnapshot();
            },
            done
        );
    });

    it('reflects appConfig properly', (done) => {
        testWithAppConfig(
            {
                storages: [
                    { path: '/path/1/', enabled: true },
                    { path: '/path/2/', enabled: false },
                ],
                ffmpeg: '/path/ffmpeg',
            },
            async ({ getByTestId }) => {
                const storage0Path = getByTestId('storage-0-path') as HTMLInputElement;
                const storage1Path = getByTestId('storage-1-path') as HTMLInputElement;
                const storage0Enabled = getByTestId('storage-0-enabled') as HTMLInputElement;
                const storage1Enabled = getByTestId('storage-1-enabled') as HTMLInputElement;
                const ffmpeg = getByTestId('ffmpeg') as HTMLInputElement;
                expect(storage0Path.value).toBe('/path/1/');
                expect(storage1Path.value).toBe('/path/2/');
                expect(storage0Enabled.checked).toBe(true);
                expect(storage1Enabled.checked).toBe(false);
                expect(ffmpeg.value).toBe('/path/ffmpeg');
            },
            done
        );
    });

    it('adds storage properly', (done) => {
        testWithAppConfig(
            {
                storages: [
                    { path: '/path/1/', enabled: true },
                    { path: '/path/2/', enabled: false },
                ],
            },
            async ({ getByTestId }) => {
                fireEvent.click(getByTestId('add-storage'));
                const storage2Path = getByTestId('storage-2-path') as HTMLInputElement;
                const storage2Enabled = getByTestId('storage-2-enabled') as HTMLInputElement;
                expect(storage2Path.value).toBe('');
                expect(storage2Enabled.checked).toBe(true);
            },
            done
        );
    });

    it('removes storage properly', (done) => {
        testWithAppConfig(
            {
                storages: [
                    { path: '/path/1/', enabled: true },
                    { path: '/path/2/', enabled: false },
                ],
            },
            async ({ getByTestId, queryByTestId }) => {
                fireEvent.click(getByTestId('storage-0-delete'));
                const storage0Path = getByTestId('storage-0-path') as HTMLInputElement;
                const storage0Enabled = getByTestId('storage-0-enabled') as HTMLInputElement;
                expect(storage0Path.value).toBe('/path/2/');
                expect(storage0Enabled.checked).toBe(false);
                expect(queryByTestId('storage-1-path')).not.toBeInTheDocument();
            },
            done
        );
    });
});
