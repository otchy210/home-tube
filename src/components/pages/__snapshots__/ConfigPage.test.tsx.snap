// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigPage renders initial state properly 1`] = `
<div>
  <p
    class="sc-bdvvtL h1 pt-3"
  >
    Config
  </p>
  <form
    class=""
  >
    <p
      class="sc-gsDKAQ h5 mt-3"
    >
      Video storage path
    </p>
    <small
      class="text-muted form-text"
    >
      Add your video storage path which has your videos.
    </small>
    <div
      class="mt-1 hstack gap-2"
    >
      <button
        class="btn btn-secondary"
        data-testid="add-storage"
        type="button"
      >
        Add video storage path
      </button>
    </div>
    <p
      class="sc-gsDKAQ h5 mt-3"
    >
      ffmpeg path
    </p>
    <small
      class="text-muted form-text"
    >
      No need to set unless you want to change it from default.
    </small>
    <input
      class="mt-1 form-control"
      data-testid="ffmpeg"
      type="text"
      value=""
    />
    <div
      class="mt-3 hstack gap-2"
    >
      <button
        class="btn btn-primary"
        data-testid="save-config"
        disabled=""
        type="button"
      >
        Save config
      </button>
      <button
        class="btn btn-link"
        data-testid="cancel"
        disabled=""
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
  <p
    class="sc-bdvvtL h1 pt-3"
  >
    Server status 
    [MockedReloadSvg onClick="() =&gt; {
        api.getServerStatus().then(setServerStatus);
    }" width="32" height="32" className="sc-dkPtRN fYeJEY" /]
  </p>
  <p
    class="sc-gsDKAQ h5 mt-3"
  >
    Searchable videos
  </p>
  <p>
    1
  </p>
  <p
    class="sc-gsDKAQ h5 mt-3"
  >
    Storages
  </p>
  <p>
    <b>
      dummyPath
    </b>
    <br />
    <span
      class="badge bg-secondary"
    >
      waiting
    </span>
     
    1
     movies found
    <br />
  </p>
  <p
    class="sc-gsDKAQ h5 mt-3"
  >
    Meta data
  </p>
  <p>
    0
     movies queued.
    <br />
  </p>
  <p
    class="sc-gsDKAQ h5 mt-3"
  >
    Thumbnails
  </p>
  <p>
    0
     movies queued.
    <br />
  </p>
  <p
    class="sc-gsDKAQ h5 mt-3"
  >
    Snapshot
  </p>
  <p>
    0
     movies queued.
    <br />
  </p>
</div>
`;
